sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java

```diff
@@ -481,12 +485,38 @@ static class UnboundedSourceAsSDFWrapperFn<OutputT, CheckpointT extends Checkpoi
     private static final Logger LOG = LoggerFactory.getLogger(UnboundedSourceAsSDFWrapperFn.class);
     private static final int DEFAULT_BUNDLE_FINALIZATION_LIMIT_MINS = 10;
     private final Coder<CheckpointT> checkpointCoder;
-    private @Nullable Cache<Object, UnboundedReader<OutputT>> cachedReaders;
+    private final MemoizingPerInstantiationSerializableSupplier<
+            Cache<Object, UnboundedReader<OutputT>>>
+        readerCacheSupplier;
+    private static final Executor closeExecutor =
+        Executors.newCachedThreadPool(
+            new ThreadFactoryBuilder().setNameFormat("UnboundedReaderCloses-%d").build());
     private @Nullable Coder<UnboundedSourceRestriction<OutputT, CheckpointT>> restrictionCoder;
 
     @VisibleForTesting
     UnboundedSourceAsSDFWrapperFn(Coder<CheckpointT> checkpointCoder) {
       this.checkpointCoder = checkpointCoder;
+      readerCacheSupplier =
+          new MemoizingPerInstantiationSerializableSupplier<>(
+              () ->
+                  CacheBuilder.newBuilder()
+                      .expireAfterWrite(1, TimeUnit.MINUTES)
+                      .removalListener(
+                          (RemovalListener<Object, UnboundedReader<OutputT>>)
+                              removalNotification -> {
+                                if (removalNotification.getCause() != RemovalCause.EXPLICIT) {
+                                  closeExecutor.execute(
+                                      () -> {
+                                        try {
+                                          Preconditions.checkNotNull(removalNotification.getValue())
```

Comment:
nit: this doesn't actually change anything, as it will just throw an NPE same as if you didn't use it. So prefer to use Beam SDK's own Preconditions which have `checkArgumentNotNull` (throws `IllegalArgumentException`) and `checkStateNotNull` (throws `IllegalStateException`)

Out of curiosity, I looked at the docs on RemovalNotification I believe if the value is null here that means it was already garbage collected. Since you aren't using `weakValues` this will never happen. But if you wanted this code to be agnostic to the cache configuration a bit, you could just check if it is null and return early.